### PR TITLE
Sprint1

### PR DESCRIPTION
**Pull request by leowe**
Implementation for Sprint 1. This includes core functionalities for Chunk, ValueSegment, Table, StorageManager.

### CODE REVIEW

File: src/lib/storage/chunk.cpp [hunk 1]
```diff
@@ -15,26 +15,26 @@
...
 void Chunk::add_segment(std::shared_ptr<BaseSegment> segment) {
-  // Implementation goes here
+  Assert(size() == 0 || segment->size() == size(), "Segment needs to have same number of rows as chunk");
+  _segments.push_back(segment);
 }
 
 void Chunk::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
+  DebugAssert(values.size() == column_count(), "New values need same number of values as chunk columns");
+  for (auto i = 0ul, end = values.size(); i < end; i++) {
```
Comment (phSch08) [inline src/lib/storage/chunk.cpp:24]: `values.size()` is evaluated in each iteration - save `values.size()` to variable and use it instead

File: src/lib/storage/chunk.cpp [hunk 2]
```diff
@@ -15,26 +15,26 @@
...
-  // Implementation goes here
-  return nullptr;
-}
+std::shared_ptr<BaseSegment> Chunk::get_segment(ColumnID column_id) const { return _segments.at(column_id); }
 
-ColumnCount Chunk::column_count() const {
-  // Implementation goes here
-  return ColumnCount{0};
-}
+ColumnCount Chunk::column_count() const { return ColumnCount{_segments.size()}; }
```
Comment (phSch08) [inline src/lib/storage/chunk.cpp:31]: Compiler throws an error: narrow conversion of long unsigned int to short unsigned int - 
```suggestion
ColumnCount Chunk::column_count() const { return ColumnCount{static_cast<ColumnCount::base_type>(_segments.size())}; }
```

File: src/lib/storage/storage_manager.cpp [hunk 3]
```diff
@@ -1,5 +1,6 @@
 #include "storage_manager.hpp"
 
+#include <iostream>
```
Comment (phSch08) [inline src/lib/storage/storage_manager.cpp:3]: already included

File: src/lib/storage/storage_manager.cpp [hunk 4]
```diff
@@ -10,38 +11,40 @@
...
-  return *(new StorageManager());
-  // A really hacky fix to get the tests to run - replace this with your implementation
+  static StorageManager storageManager;
+  return storageManager;
 }
 
 void StorageManager::add_table(const std::string& name, std::shared_ptr<Table> table) {
-  // Implementation goes here
+  Assert(!has_table(name), "Cannot add table, because a table with this name already exists");
+  _tables[name] = table;
```
Comment (phSch08) [inline src/lib/storage/storage_manager.cpp:20]: Use `_tables.insert()` with return value - see lecture

File: src/lib/storage/storage_manager.cpp [hunk 5]
```diff
@@ -10,38 +11,40 @@
...
 }
 
 void StorageManager::drop_table(const std::string& name) {
-  // Implementation goes here
+  Assert(has_table(name), "Cannot drop non-existing table");
+  _tables.erase(name);
```
Comment (phSch08) [inline src/lib/storage/storage_manager.cpp:25]: Use `_tables.erase()` with return value - see lecture

File: src/lib/storage/storage_manager.cpp [hunk 6]
```diff
@@ -10,38 +11,40 @@
...
 
-bool StorageManager::has_table(const std::string& name) const {
-  // Implementation goes here
-  return false;
-}
+bool StorageManager::has_table(const std::string& name) const { return _tables.contains(name); }
 
 std::vector<std::string> StorageManager::table_names() const {
-  throw std::runtime_error("Implement StorageManager::table_names");
+  auto table_names = std::vector<std::string>();
```
Comment (phSch08) [inline src/lib/storage/storage_manager.cpp:33]: Since we know the size here, use `table_names.reserve()`

File: src/lib/storage/table.cpp [hunk 7]
```diff
@@ -18,56 +18,65 @@
...
+  const auto completed_chunks_size = (chunk_count() - 1) * _target_chunk_size;
+  const auto last_chunk_size = _chunks[chunk_count() - 1]->size();
+  return completed_chunks_size + last_chunk_size;
 }
 
-ChunkID Table::chunk_count() const {
-  // Implementation goes here
-  return ChunkID{0};
-}
+ChunkID Table::chunk_count() const { return ChunkID{_chunks.size()}; }
```
Comment (phSch08) [inline src/lib/storage/table.cpp:53]: Compiler Error: narrowing conversion of long unsigned int to uint32_t (unsinged int)
```suggestion
ChunkID Table::chunk_count() const { return ColumnCount{static_cast<ColumnCount::base_type>(_column_names.size())}; }
```

File: src/lib/storage/table.cpp [hunk 8]
```diff
@@ -18,56 +18,65 @@
...
-  return ColumnCount{0};
-}
+ColumnCount Table::column_count() const { return _chunks[0]->column_count(); }
 
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  return 0;
+  const auto completed_chunks_size = (chunk_count() - 1) * _target_chunk_size;
+  const auto last_chunk_size = _chunks[chunk_count() - 1]->size();
+  return completed_chunks_size + last_chunk_size;
```
Comment (phSch08) [inline src/lib/storage/table.cpp:50]: chunks can have different sizes in the future

File: src/lib/storage/table.cpp [hunk 9]
```diff
@@ -18,56 +18,65 @@
...
+  _column_names.push_back(name);
+  _column_types.push_back(type);
+  _add_value_segment_to_chunk(_chunks[0], type);
 }
 
 void Table::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
+  auto last_chunk = _chunks[chunk_count() - 1];
+  if (last_chunk->size() == _target_chunk_size) {
+    last_chunk = std::make_shared<Chunk>();
```
Comment (phSch08) [inline src/lib/storage/table.cpp:36]: Use a new variable for improved readability

File: src/lib/storage/chunk.cpp (hunk 1 above)
Comment (mweisgut) [inline src/lib/storage/chunk.cpp:24]: Please choose clear and concise names, and avoid, e.g., i, j, ch_ptr.

File: src/lib/storage/storage_manager.cpp [hunk 10]
```diff
@@ -10,38 +11,40 @@
...
+  for (const auto& [table_name, table] : _tables) {
```
Comment (Bensk1) [inline src/lib/storage/storage_manager.cpp:34]: Nice use of structured bindings.

File: src/lib/storage/table.cpp [hunk 11]
```diff
@@ -18,56 +18,65 @@
...
 
 Table::Table(const ChunkOffset target_chunk_size) {
-  // Implementation goes here
+  _target_chunk_size = target_chunk_size;
+  _chunks.push_back(std::make_shared<Chunk>());
 }
 
 void Table::add_column(const std::string& name, const std::string& type) {
-  // Implementation goes here
+  Assert(row_count() == 0, "Cannot add column to non-empty table");
```
Comment (Bensk1) [inline src/lib/storage/table.cpp:26]: Good!

File: src/lib/storage/chunk.cpp [hunk 12]
```diff
@@ -15,26 +15,26 @@
...
-  // Implementation goes here
+  Assert(size() == 0 || segment->size() == size(), "Segment needs to have same number of rows as chunk");
+  _segments.push_back(segment);
 }
 
 void Chunk::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
+  DebugAssert(values.size() == column_count(), "New values need same number of values as chunk columns");
+  for (auto i = 0ul, end = values.size(); i < end; i++) {
+    _segments[i]->append(values[i]);
```
Comment (mweisgut) [inline src/lib/storage/chunk.cpp:25]: more descriptive identifier names, see above

File: src/lib/storage/table.cpp [hunk 13]
```diff
@@ -18,56 +18,65 @@
...
+
+  _column_names.push_back(name);
+  _column_types.push_back(type);
+  _add_value_segment_to_chunk(_chunks[0], type);
 }
 
 void Table::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
+  auto last_chunk = _chunks[chunk_count() - 1];
```
Comment (Bensk1) [inline src/lib/storage/table.cpp:34]: Could also be `_chunks.back()`

File: src/lib/storage/table.cpp [hunk 14]
```diff
@@ -18,56 +18,65 @@
...
-  // Implementation goes here
-  return ChunkID{0};
-}
+ChunkID Table::chunk_count() const { return ChunkID{_chunks.size()}; }
 
 ColumnID Table::column_id_by_name(const std::string& column_name) const {
-  // Implementation goes here
-  return ColumnID{0};
-}
+  const auto it = std::find(_column_names.begin(), _column_names.end(), column_name);
```
Comment (Bensk1) [inline src/lib/storage/table.cpp:56]: Minor: Could be `cbegin()` and `cend()`.

**Pull request closed without merging**